src/cmds/scripts/pbs_habitat.in

```diff
@@ -246,9 +246,17 @@ upgrade_pbs_database() {
 	[ ${sys_pgsql_ver%.*} -eq ${old_pgsql_ver%.*} ] && [ ${sys_pgsql_ver#*.} \> ${old_pgsql_ver#*.} ] || [ ${sys_pgsql_ver%.*} -gt ${old_pgsql_ver%.*} ];
 	result=$?
 	if [ ${result} -eq 0 ]; then
-		#Start upgrade process of datastore
-		upgrade_db
-		return $?
+		if [ -d "$PBS_EXEC/pgsql" ]; then
+			#Start upgrade process of datastore
+			upgrade_db
+			return $?
+		else
+			echo "postgresql upgraded without upgrading PBS database."
+			echo "PBS database needs to be upgraded."
+			echo "Please follow the upgrade instructions in the release notes for the v18.1.2 release at:"
+			echo "https://github.com/PBSPro/pbspro/releases"
+			exit 1
```

Comment:
You don't want to "exit 1", you want to "return 1".
For the message I would suggest:

It appears that PostgreSQL has been upgraded independently of PBS Pro.
The PBS Pro database must be manually upgraded. Please refer to the
instructions in the release notes for version @PBS_VERSION@ located here:
https://github.com/PBSPro/pbspro/releases

Autotools will take care of substituting @PBS_VERSION@ for you.

Reply:
Line lengths in the suggested text were intentional.

Reply:
If I use return 1, we will be displaying "Failed to upgrade PBS Datastore". I will change the text as you have suggested.

Reply:
Perhaps we need to return 2 and handle that case with a different message like... PBS Pro database upgrade could not be performed automatically.

Reply:
My point is that there are additional instructions that may need to be run AFTER the databse upgrade. Exiting the script would prevent them from running.

Reply:
but, we are not upgrading the database when we see a version mismatch, so exit looks ok to me unless I am missing anything...

Reply:
I have updated the EDD and code to use the new message. Could you please have a look.
Thanks,
Prakash

Reply:
I feel rather strongly against using "exit 1" because (at some point in the future) someone might add code to pbs_habitat after the call to upgrade the DB. In some cases, it will never run and in others it will though the intent was for the new code to always run.

Feel free to remove the @PBS_VERSION@ and simply reference the release notes. The admin should know which version they are installing. In this case, we were already working with a ".in" file so it was easy to add.

Reply:
I would like to keep @PBS_VERSION@ for now.
Coming to "exit 1", at this time, when we see a version mismatch, it makes sense to not display anything apart from what we are already displaying. In future, if there is code to be run even after version mismatch is detected, we can change exit 1 to return 1, or 2, whatever is appropriate at that time.

Reply:
The problem is that developers are unlikely to notice that there is an "exit 1" buried in a function when they add code to the end of the script. Please return 2 and handle it gracefully in the caller.

Reply:
We're returning in case of failure in all other failure scenarios within `upgrade_pbs_database` and it's caller's responsibility to handle them. It's better to follow the same convention. Rest of the changes looks good to me!

Reply:
@mike0042, @nithinj - even though I do not feel the change was necessary, I have changed exit to return. Please have a look.

Reply:
I now see your point about not wanting to change the call to exit, since we end up exiting in the caller. I also think that's the wrong thing to do, but it appears consistent with the way the script is written. Something we'll end up refactoring someday.